src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityAssemblyLine.java

```diff
@@ -60,4 +69,33 @@ public ICubeRenderer getBaseTexture(IMultiblockPart sourcePart) {
     protected IBlockState getCasingState() {
         return MetaBlocks.METAL_CASING.getState(BlockMetalCasing.MetalCasingType.STEEL_SOLID);
     }
+
+    @Override
+    public boolean checkRecipe(Recipe recipe, boolean consumeIfSuccess) {
+        if (recipe.hasProperty(ResearchItemProperty.getInstance())) {
+            List<IItemHandlerModifiable> dataStickHandler = this.getAbilities(MultiblockAbility.RESEARCH_DATA);
+            List<ItemStack> validOutputs = new ArrayList<>();
+            if (!dataStickHandler.isEmpty()) {
+                List<ItemStack> hatchDataList = GTUtility.itemHandlerToList(dataStickHandler.get(0));
+                for (ItemStack stack : hatchDataList) {
+                    NBTTagCompound researchItemNBT = stack.getSubCompound("asslineOutput");
+                    if (researchItemNBT != null) {
+                        validOutputs.add(new ItemStack(researchItemNBT));
+                    }
+                }
+            }
+            boolean canRun = false;
+            for (ItemStack stack: validOutputs) {
+                String stackStr = stack.toString();
+                String recipePropertyString = recipe.getOutputs().get(0).toString();
+                if (stackStr.equals(recipePropertyString)) {
+                    canRun = true;
+                    break;
+                }
+            }
+            return canRun;
```

Comment:
Instead of getting all the data sticks' items and then checking those items afterwards again, you should be able to do this:
```suggestion
                    if (researchItemNBT != null && new ItemStack(researchItemNBT).isItemEqual(recipe.getOutputs().get(0)))
                            return true;
                }
            }
            return false;
```